mijia/src/lib.rs

```diff
@@ -17,85 +18,127 @@ const CONNECTION_INTERVAL_CHARACTERISTIC_PATH: &str = "/service0021/char0045";
 const CONNECTION_INTERVAL_500_MS: [u8; 3] = [0xF4, 0x01, 0x00];
 const DBUS_METHOD_CALL_TIMEOUT: Duration = Duration::from_secs(30);
 
+#[derive(Clone, Debug)]
 pub struct SensorProps {
-    pub object_path: String,
-    pub mac_address: String,
+    pub id: DeviceId,
+    pub mac_address: MacAddress,
 }
 
-pub async fn get_sensors(bt_session: &MijiaSession) -> Result<Vec<SensorProps>, anyhow::Error> {
-    let bluez_root = dbus::nonblock::Proxy::new(
-        "org.bluez",
-        "/",
-        DBUS_METHOD_CALL_TIMEOUT,
-        bt_session.connection.clone(),
-    );
-    let tree = bluez_root.get_managed_objects().await?;
-
-    let sensors = tree
-        .into_iter()
-        .filter_map(|(path, interfaces)| {
-            // FIXME: can we generate a strongly typed deserialiser for this,
-            // based on the introspection data?
-            let device_properties = interfaces.get("org.bluez.Device1")?;
-
-            let mac_address = device_properties
-                .get("Address")?
-                .as_iter()?
-                .filter_map(|addr| addr.as_str())
-                .next()?
-                .to_string();
-            // UUIDs don't get populated until we connect. Use:
-            //     "ServiceData": Variant(InternalDict { data: [
-            //         ("0000fe95-0000-1000-8000-00805f9b34fb", Variant([48, 88, 91, 5, 1, 23, 33, 215, 56, 193, 164, 40, 1, 0])
-            //     )], outer_sig: Signature("a{sv}") })
-            // instead.
-            let service_data: HashMap<String, _> = device_properties
-                .get("ServiceData")?
-                // Variant(...)
-                .as_iter()?
-                .next()?
-                // InternalDict(...)
-                .as_iter()?
-                .tuples::<(_, _)>()
-                .filter_map(|(k, v)| Some((k.as_str()?.into(), v)))
-                .collect();
+// TODO before publishing to crates.io: annotate this enum as non-exhaustive.
+/// An event from a Mijia sensor.
+#[derive(Clone)]
+pub enum MijiaEvent {
+    // FIXME: stop using object_path as primary key. Can we think of something better?
+    Readings { id: DeviceId, readings: Readings },
+    Disconnected { id: DeviceId },
+}
 
-            if service_data.contains_key(MIJIA_SERVICE_DATA_UUID) {
-                Some(SensorProps {
-                    object_path: path.to_string(),
-                    mac_address,
+impl MijiaEvent {
+    fn from(conn_msg: Message) -> Option<Self> {
+        match BluetoothEvent::from(conn_msg) {
+            Some(BluetoothEvent::Value { object_path, value }) => {
+                // TODO: Make this less hacky.
+                let object_path = object_path
+                    .strip_suffix(SENSOR_READING_CHARACTERISTIC_PATH)?
+                    .to_owned();
+                let readings = Readings::decode(&value)?;
+                Some(MijiaEvent::Readings {
+                    id: DeviceId { object_path },
```

Comment:
maybe make a `DeviceId::new(&str)` and do the clone in there?

Reply:
Done.